runtime/lua/vim/iter.lua

```diff
@@ -179,6 +194,51 @@ function ListIter.filter(self, f)
   return self
 end
 
+--- Flattens an |list-iterator| pipeline, un-nesting nested values.
+--- The flattened list will exclude nil, and any values with non-numeric keys
+---
+--- Example:
+---
+--- ```lua
+--- vim.iter({1, {2}, {{3}}}):flat():totable()
+--- ```
+--- (returns `{1, 2, {3}}`)
+---
+--- ```lua
+--- vim.iter({1, {a: 2}, {{3}}}):flat():totable()
+--- ```
+--- (returns `{1, 3}`)
+---
+---@param depth number|nil Depth to which iterator pipeline should be flattened
+---                        (defaults to 1)
+---@return Iter
+function Iter.flat(self, depth) -- luacheck: no unused args
+  error('flat() requires a list-like table')
+  return self
+end
+
+---@private
+function ListIter.flat(self, depth)
+  depth = depth or 1
+  local inc = self._head < self._tail and 1 or -1
+  local target = {}
+
+  local n = 1
+  for i = self._head, self._tail - inc, inc do
+    local flattened = flatten(self._table[i], depth)
+    if flattened ~= nil then
+      for _, flat_v in pairs(flattened) do
```

Comment:
pairs or ipairs?

Reply:
good call, should be `ipairs`. We shouldn't have any non-numeric keys at this point, but that is safer and makes it more clear what we're doing

Reply:
Also, the behaviors on a table like `{ nil, 2 }` or `{ [2] = 2 }` are different. It would be great if this can be included in the tests.

Reply:
hmm. not allowing `nil` in a list-like iterator seems to be a peculiarity of `Iter:new()`. The second case makes more sense, because that table isn't really list-like. I can add tests and examples in the docs for these

Reply:
lists are tables that are indexed with consecutive(!) integers starting from 1, so the first table is also not list-like.

Reply:
@clason interesting, I would have thought the structure was `{[1] = nil, [2] = 2}`, did not know lua works like this:

```lua
lua print(vim.inspect({nil, 2}))
```
gives `{ [2] = 2 }`

Reply:
You use `nil` to delete a key, so these two tables are (have to be) equivalent.